runtime/protocol/http/common/src/aws/smithy/kotlin/runtime/http/middleware/FlexibleChecksumsRequestMiddleware.kt

```diff
@@ -0,0 +1,117 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package aws.smithy.kotlin.runtime.http.middleware
+
+import aws.smithy.kotlin.runtime.ClientException
+import aws.smithy.kotlin.runtime.hashing.*
+import aws.smithy.kotlin.runtime.http.*
+import aws.smithy.kotlin.runtime.http.operation.*
+import aws.smithy.kotlin.runtime.http.request.header
+import aws.smithy.kotlin.runtime.io.*
+import aws.smithy.kotlin.runtime.util.*
+import kotlinx.coroutines.*
+import kotlin.coroutines.coroutineContext
+
+/**
+ * Mutate a request to enable flexible checksums.
+ *
+ * If the checksum will be sent as a header, calculate the checksum.
+ *
+ * Otherwise, if it will be sent as a trailing header, calculate the checksum as asynchronously as the body is streamed.
+ * In this case, a [LazyAsyncValue] will be added to the execution context which allows the trailing checksum to be sent
+ * after the entire body has been streamed.
+ *
+ * @param checksumAlgorithmName the name of the algorithm used to calculate the checksum
+ */
+@InternalApi
+public class FlexibleChecksumsRequestMiddleware(private val checksumAlgorithmName: String) : ModifyRequestMiddleware {
+    public override suspend fun modifyRequest(req: SdkHttpRequest): SdkHttpRequest {
+        val logger = coroutineContext.getLogger<FlexibleChecksumsRequestMiddleware>()
+
+        val headerName = "x-amz-checksum-${checksumAlgorithmName.lowercase()}"
+        logger.debug { "Resolved header name: $headerName" }
+
+        // remove all checksum headers except for $headerName
+        // this handles the case where a user inputs a precalculated checksum, but it doesn't match the input checksum algorithm
+        req.subject.headers.removeAllChecksumHeadersExcept(headerName)
+
+        val checksumAlgorithm = checksumAlgorithmName.toHashFunction() ?: throw ClientException("Could not parse checksum algorithm $checksumAlgorithmName")
+
+        if (!checksumAlgorithm.isSupported) {
```

Comment:
**Question**: if we've converted a checksum algorithm name to an actual runtime type when would this ever not be true?

Reply:
`.toHashFunction()` also supports parsing MD5 (intended for other usecases like HashingSource/ByteReadChannel), so if the user modifies the `ChecksumAlgorithm` enum to pass "md5", it will convert to the proper runtime type, but MD5 is not a supported algorithm for flexible checksums.

In general this issue arises because `.toHashFunction()` was intended to be used for more than just flexible checksums. I could solve it by making an `internal` function which does the exact same thing, without MD5. What do you think?